### PR TITLE
perf: Optimize time-based filtering on large number of workflows

### PR DESCRIPTION
**Pull request by terrytangyuan**
This PR introduces a new `WorkflowPredicate` for filtering workflows based on a time range, which brings performance gains when listing large number of workflows, especially when the first filtering condition only filtered out very small number of workflows this could be a huge speed-up.

Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -125,6 +125,11 @@ var (
 			return !wf.Status.FinishedAt.IsZero() && wf.Status.FinishedAt.Time.Before(t)
 		}
 	}
+	WorkflowHappenedBetween = func(startTime time.Time, endTime time.Time) WorkflowPredicate {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:128]: WorkflowRanBetween?

Comment (terrytangyuan) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Much better. Thanks!

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -125,6 +125,11 @@ var (
 			return !wf.Status.FinishedAt.IsZero() && wf.Status.FinishedAt.Time.Before(t)
 		}
 	}
+	WorkflowRanBetween = func(startTime time.Time, endTime time.Time) WorkflowPredicate {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:128]: interesting, I can think of other interpretation of this:

* the workflow started between times A and B
* the workflow finished between times A and. B
* the workflow was running between the two times - i.e. start time is before B and (not finished OR finished after time A)

Comment (terrytangyuan) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:128]: Yes. Should we discuss further on a better design in a separate issue? The original purpose of this PR is to optimize the performance of the existing functionality. We can definitely break this down and see if we can come up something more comprehensive and useful in practice.

**Pull request merged**